### PR TITLE
Port the tag stripping memcpy detection to RISC-V

### PR DESCRIPTION
**Pull request by arichardson**
This debugging features has been extremely useful in the past when
debugging tag losses in complex software. I made one minor change compared
to jdw57_memcpy: it can also be configured by setting the environment
variable CHERI_ABORT_ON_TAG_STRIPPING_COPY to 0/1 rather than being
restricted to a root-only global sysctl.

### CODE REVIEW

**Comment by jrtc27** [comment]
I don't like this, and view it as a feature not a bug that we ditched all the bogus debugging crud from jdw57_memcpy.c (both the abort and the warning message). It is fundamentally broken given that on-stack buffers contain leftover junk from prior usage that may be tagged and so gives false-positives, and slows down memcpy. I never want it to be compiled into my builds, let alone benchmarking ones.

**Comment by jrtc27** [comment]
But regardless, this belongs in a common C file or header that jdw57_memcpy can share with it, as well as to keep bcopy.c as untainted as possible.

File: bin/cheritest/cheritest.c [hunk 1]
```diff
@@ -1647,12 +1647,16 @@ static const struct cheri_test cheri_tests[] = {
...
+	  .ct_flags = CT_FLAG_SIGEXIT,
+	  .ct_signum = SIGABRT },
 	{ .ct_name = "test_unaligned_capability_copy_memmove",
 	  .ct_desc = "Check that a memmove() of valid capabilities to an "
-		     "unaligned destination strips tags",
-	  .ct_func = test_unaligned_capability_copy_memmove },
+		     "unaligned destination fails",
+	  .ct_func = test_unaligned_capability_copy_memmove,
+	  .ct_flags = CT_FLAG_SIGEXIT,
+	  .ct_signum = SIGABRT },
```
Comment (jrtc27) [inline bin/cheritest/cheritest.c:1659]: Add tests, don't change the existing valid ones.

File: lib/libc/string/bcopy.c [hunk 2]
```diff
@@ -65,6 +71,104 @@ typedef	int word;		/* "word" used for optimal copy speed */
...
+		size_t olen = sizeof(abort_on_tag_loss);
+		if (__sysctlbyname(ABORT_ON_TAG_LOSS_SYSCTL,
+			strlen(ABORT_ON_TAG_LOSS_SYSCTL), &abort_on_tag_loss,
+			&olen, NULL, 0) == -1) {
+			write_to_stderr("ERROR: could not determine whether "
+			    "tag stripping memcpy should abort. Assuming it "
+			    "shouldn't.\n");
+			abort_on_tag_loss = 0;
+		}
+	}
```
Comment (jrtc27) [inline lib/libc/string/bcopy.c:122]: The global sysctl makes no sense, on CHERI-MIPS I always see a few in normal use so that would make processes crash. If you insist on having it have it as an environment variable only, otherwise you're giving people an option to shoot themselves in the foot (and they _will_ as it's confusingly in the "security" namespace).

**Comment by jrtc27** [comment]
Also, this won't help you if/when we add hardware support for unaligned capability memcpy, since those _must_ silently strip tags and so you won't hit the memcpy case.

File: lib/libc/string/bcopy.c (hunk 2 above)
Comment (arichardson) [reply lib/libc/string/bcopy.c:121]: Those are bugs that we haven't bothered fixing (since it no longer abort()s). Why would copying unitialized data from the stack be useful? I get it that some people believe it's a source of entropy but please just use a tag-stripping memcpy().

File: bin/cheritest/cheritest.c (hunk 1 above)
Comment (arichardson) [reply bin/cheritest/cheritest.c:1659]: I'm restoring the old behaviour.

**Comment by arichardson** [comment]
> But regardless, this belongs in a common C file or header that jdw57_memcpy can share with it, as well as to keep bcopy.c as untainted as possible.

I think jdw57_memcpy.c should be deleted, we can get the same performance if we compile the file without debug info. Or we just accept the extra nop in the delay slot.

**Comment by arichardson** [comment]
> I don't like this, and view it as a feature not a bug that we ditched all the bogus debugging crud from jdw57_memcpy.c (both the abort and the warning message). It is fundamentally broken given that on-stack buffers contain leftover junk from prior usage that may be tagged and so gives false-positives, and slows down memcpy. I never want it to be compiled into my builds, let alone benchmarking ones.

This only slows down memcpy() that is doing a byte-by-byte copy and is therefore slow already. And yes it should be under #ifndef NDEBUG

File: lib/libc/string/bcopy.c (hunk 2 above)
Comment (jrtc27) [reply lib/libc/string/bcopy.c:121]:
```
struct S {
    int flag;
    char buf[BUF_LEN];
};

void foo(char *str) {
    struct S s;
    s.flag = 1;
    strcpy(s.buf, str);
    bar(&s);
}

void bar(struct S *s) {
    char buf[BUF_LEN];
    memcpy(buf, s->buf);
    baz(buf);
}
```

Bit contrived but I hope you can see how there are legitimate coding patterns that can end up copying uninitialised stack data.

Comment (jrtc27) [reply lib/libc/string/bcopy.c:121]: Or, worse, if unions are involved you can have a _lot_ of padding that's uninitialised garbage.

Comment (arichardson) [reply lib/libc/string/bcopy.c:121]: The union case I agree could be an issue (although C programmers very rarely pass those by value / copy them). Do you know what MSAN does there? Maybe opt-out of instrumenting those copies?

Comment (jrtc27) [reply lib/libc/string/bcopy.c:121]: I don't know, nor do I particularly care, because I believe it is fundamentally not memcpy's problem to solve.

**Comment by jrtc27** [comment]
Needs redoing from scratch if at all, no point keeping this PR open
